File: tools/github/cmake.py
# Copyright (c) 2023 Marcin Zdun
# This code is licensed under MIT license (see LICENSE for details)

import os
import re
from typing import Iterator, List, NamedTuple, Optional

PROJECT_SOURCE_DIRECTORY = os.path.dirname(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
)

TOKENS = [
    ("COMMENT", r"#.*"),
    ("STR", r'"[^"]*"'),
    ("OPEN", r"\("),
    ("CLOSE", r"\)"),
    ("IDENT", r'[^ \t\r\n()#"]+'),
    ("WS", r"\s+"),
]


class Token(NamedTuple):
    type: str
    value: str
    offset: int


def _token_stream(text: str) -> Iterator[Token]:
    tok_regex = "|".join("(?P<%s>%s)" % pair for pair in TOKENS)
    get_token = re.compile(tok_regex).match
    offset = 0

    mo = get_token(text)
    while mo is not None:
        token_type = mo.lastgroup
        if token_type not in ["COMMENT", "WS"]:
            value = mo.group(token_type)
            yield Token(type=token_type, value=value, offset=offset)
        offset = mo.end()
        mo = get_token(text, offset)


class Arg(NamedTuple):
    value: str
    offset: int


class Command(NamedTuple):
    name: str
    args: List[Arg]
    offset: int


def _command(cmd: Token, stream: Iterator[Token]):
    result = Command(name=cmd.value, args=[], offset=cmd.offset)
    if next(stream)[0] != "OPEN":
        return result
    for tok in stream:
        if tok.type == "CLOSE":
            break
        elif tok.type == "OPEN":
            continue
        elif tok.type == "STR":
            result.args.append(Arg(value=tok.value[1:-1], offset=tok.offset + 1))
        elif tok.type == "IDENT":
            result.args.append(Arg(value=tok.value, offset=tok.offset))
        else:
            print(tok)

    return result


def _cmake(filename: str):
    commands: List[Command] = []

    with open(filename, "r", encoding="UTF-8") as f:
        stream = _token_stream(f.read())
    for tok in stream:
        if tok.type == "IDENT":
            commands.append(_command(tok, stream))

    return commands


class Project(NamedTuple):
    name: Arg
    version: Arg
    stability: Arg
    description: Arg

    def ver(self):
        return f"{self.version.value}{self.stability.value}"

    def pkg(self):
        return f"{self.name.value}-{self.ver()}"

    def tag(self):
        return f"v{self.ver()}"


_project_version: Project = None


def get_version() -> Project:
    global _project_version
    if _project_version is not None:
        return _project_version
    commands = _cmake(os.path.join(PROJECT_SOURCE_DIRECTORY, "CMakeLists.txt"))

    project_name: Optional[Arg] = None
    version = Arg("0.1.0", -1)
    version_stability: Optional[Arg] = None
    description = Arg("", -1)

    for cmd in commands:
        if cmd.name == "project":
            project_name = cmd.args[0]
            args = cmd.args[1:]
            project_dict = {
                name.value: value for name, value in zip(args[::2], args[1::2])
            }
            version = project_dict.get("VERSION", version)
            description = project_dict.get("DESCRIPTION", "")
            if version_stability is not None:
                break
        elif cmd.name == "set":
            var_name = cmd.args[0]
            if var_name.value == "PROJECT_VERSION_STABILITY":
                version_stability = cmd.args[1]
                if project_name is not None:
                    break

    if project_name is None:
        project_name = Arg("", -1)
    if version_stability is None:
        version_stability = Arg("", -1)

    _project_version = Project(
        name=project_name,
        version=version,
        stability=version_stability,
        description=description,
    )

    return _project_version


def _patch(arg: Arg, value: str):
    global _project_version

    with open(
        os.path.join(PROJECT_SOURCE_DIRECTORY, "CMakeLists.txt"), "r", encoding="UTF-8"
    ) as input:
        text = input.read()

    patched = text[: arg.offset] + value + text[arg.offset + len(arg.value) :]

    with open(
        os.path.join(PROJECT_SOURCE_DIRECTORY, "CMakeLists.txt"), "w", encoding="UTF-8"
    ) as input:
        input.write(patched)

    _project_version = None


def set_version(ver: str):
    ver_split = ver.split("-", 1)
    ver = ver_split[0]

    _patch(get_version().version, ver)
    if len(ver_split) > 1:
        stability = f"-{ver_split[1]}"
        _patch(get_version().stability, stability)
    elif len(get_version().stability.value):
        _patch(get_version().stability, "")
        

File: src/base/cmake.cc
// Copyright (c) 2023 Marcin Zdun
// This code is licensed under MIT license (see LICENSE for details)

#include "base/cmake.hh"
#include <optional>
#include <span>
#include "io/file.hh"

#include <fmt/format.h>

namespace fs = std::filesystem;
using namespace std::literals;

namespace cmake {
	enum class tok { STR, OPEN, CLOSE, IDENT };

	struct token {
		tok type;
		std::string_view::iterator reset;
		std::string_view value{};
	};

	struct command {
		std::string name{};
		std::vector<std::string> args{};
	};

	unsigned char uchar(char c) { return c; }

	void skip_ws(std::string_view::iterator& it,
	             std::string_view::iterator end) {
		while (it != end) {
			while (it != end && std::isspace(uchar(*it)))
				++it;
			if (it == end) continue;
			if (*it != '#') break;
			while (it != end && *it != '\n')
				++it;
		}
	}

	struct token_stream {
		std::string_view::iterator it;
		std::string_view::iterator end;
		token_stream(std::string_view file)
		    : it{file.begin()}, end{file.end()} {}

		void put_back(std::optional<token> const& tok) {
			if (tok) it = tok->reset;
		}
		std::optional<token> next() {
			while (it != end) {
				skip_ws(it, end);
				if (it == end) continue;
				if (*it == '(') {
					auto reset = it;
					++it;
					return token{.type = tok::OPEN, .reset = reset};
				}
				if (*it == ')') {
					auto reset = it;
					++it;
					return token{.type = tok::CLOSE, .reset = reset};
				}
				if (*it == '"') {
					auto reset = it;
					++it;
					auto str_start = it;
					while (it != end && *it != '"')
						++it;
					auto str_end = it;
					if (it != end) ++it;
					return token{.type = tok::STR,
					             .reset = reset,
					             .value{str_start, str_end}};
				}

				auto ident_start = it;
				while (it != end && *it != '"' && *it != '(' && *it != ')' &&
				       *it != '#' && !std::isspace(uchar(*it)))
					++it;
				auto ident_end = it;
				return token{.type = tok::IDENT,
				             .reset = ident_start,
				             .value{ident_start, ident_end}};
			}

			return std::nullopt;
		}
	};

	struct command_stream {
		std::vector<std::byte> bytes;
		token_stream tokens{std::string_view{
		    reinterpret_cast<char const*>(bytes.data()), bytes.size()}};

		explicit command_stream(fs::path const& filename)
		    : bytes{open_file(filename)} {}

		std::optional<command> next() {
			while (auto next = tokens.next()) {
				if (next->type == tok::IDENT) return get_command(next->value);
			}
			return std::nullopt;
		}

	private:
		static std::vector<std::byte> open_file(fs::path const& filename) {
			auto file = io::fopen(filename);
			if (!file) return {};

			return file.read();
		}

		command get_command(std::string_view name) {
			command result{};
			result.name.assign(name);
			auto next = tokens.next();
			if (!next || next->type != tok::OPEN) {
				tokens.put_back(next);
				return result;
			}
			while (next = tokens.next()) {
				bool done = false;
				switch (next->type) {
					case tok::CLOSE:
						done = true;
						break;
					case tok::STR:
					case tok::IDENT:
						result.args.push_back(
						    {next->value.data(), next->value.size()});
				}
				if (done) break;
			}
			return result;
		}
	};

	project load_project() {
		std::optional<std::string> project_name;
		std::optional<std::string> version_stability;
		auto version = "0.1.0"s;
		auto description = ""s;

		command_stream commands{"CMakeLists.txt"};
		while (auto cmd = commands.next()) {
			if (cmd->name == "project") {
				project_name = cmd->args[0];
				auto args = std::span{cmd->args}.subspan(1);
				std::optional<std::string_view> var{std::nullopt};
				for (auto const& arg : args) {
					if (!var) {
						var = std::string_view{arg};
						continue;
					}
					auto var_name = *var;
					var = std::nullopt;
					if (var_name == "DESCRIPITON"sv) {
						description.assign(arg);
					} else if (var_name == "VERSION"sv) {
						version.assign(arg);
					}
				}
				if (version_stability) break;
				continue;
			}
			if (cmd->name == "set") {
				auto const& var_name = cmd->args[0];
				if (var_name == "PROJECT_VERSION_STABILITY"sv) {
					version_stability = cmd->args[1];
					if (project_name) break;
				}
			}
		}

		if (!project_name) project_name = ""s;
		if (!version_stability) version_stability = ""s;

		return {.name = std::move(*project_name),
		        .version = std::move(version),
		        .stability = std::move(*version_stability),
		        .description = std::move(description)};
	}

	project const& get_project() {
		static project pro{load_project()};
		return pro;
	}
}  // namespace cmake
